### PR TITLE
Add rpc command to validate the store contents

### PR DESCRIPTION
**Pull request by rklaehn**
Needed becuause I broke my store. I guess other people will do as well. It is easy to do with the files not being readonly.

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -392,6 +399,23 @@ async fn main_impl() -> Result<()> {
...
+            while let Some(item) = response.next().await {
+                let item = item?;
+                println!(
+                    "{} {} {} {}",
+                    item.path
+                        .map(|x| x.display().to_string())
+                        .unwrap_or("Collection".to_owned()),
+                    Blake3Cid(item.hash),
+                    HumanBytes(item.size),
+                    item.error.unwrap_or("OK".to_owned()),
```
Comment (dignifiedquire) [inline src/main.rs:414]: Would be nice to have like a checkmark and a cross for each state

File: src/provider/mod.rs [hunk 2]
```diff
@@ -404,6 +425,23 @@ impl RpcHandler {
...
     }
+
+    /// Invoke validate on the database and stream out the result
+    fn validate(
+        self,
+        _msg: ValidateRequest,
+    ) -> impl Stream<Item = ValidateResponse> + Send + 'static {
+        self.inner
+            .db
+            .validate(4)
```
Comment (dignifiedquire) [inline src/provider/mod.rs:436]: constant, to make clear what this number means

File: src/util.rs [hunk 3]
```diff
@@ -152,6 +154,46 @@ impl From<anyhow::Error> for RpcError {
...
+) -> result::Result<(), BaoValidationError> {
+    let hash = blake3::Hash::from(hash);
+    let outboard_reader = io::Cursor::new(outboard);
+    let mut decoder = abao::decode::Decoder::new_outboard(data_reader, outboard_reader, &hash);
+    // todo: expose chunk group size in abao, so people can allocate good sized buffers
+    let mut buffer = [0u8; 1024 * 16 + 4096];
+    loop {
+        match decoder.read(&mut buffer) {
+            Ok(0) => break,
+            Ok(_) => {}
```
Comment (dignifiedquire) [inline src/util.rs:187]: shouldn't you make sure the whole thing gets read, by summing the read amount?

File: src/util.rs [hunk 4]
```diff
@@ -152,6 +154,46 @@ impl From<anyhow::Error> for RpcError {
...
+pub(crate) fn validate_bao(
+    hash: Hash,
+    data_reader: impl Read + Seek,
+    outboard: Bytes,
+) -> result::Result<(), BaoValidationError> {
+    let hash = blake3::Hash::from(hash);
+    let outboard_reader = io::Cursor::new(outboard);
+    let mut decoder = abao::decode::Decoder::new_outboard(data_reader, outboard_reader, &hash);
+    // todo: expose chunk group size in abao, so people can allocate good sized buffers
+    let mut buffer = [0u8; 1024 * 16 + 4096];
```
Comment (dignifiedquire) [inline src/util.rs:183]: this is quite large for a stack array, probably makes more sense to use a Vec or BytesMut

File: src/main.rs [hunk 5]
```diff
@@ -392,6 +399,23 @@ async fn main_impl() -> Result<()> {
             }
             Ok(())
         }
+        Commands::Validate { rpc_port } => {
+            let client = make_rpc_client(rpc_port).await?;
+            let mut response = client.server_streaming(ValidateRequest).await?;
+            while let Some(item) = response.next().await {
```
[the remark this answers is not in the document]
Comment (rklaehn) [reply src/main.rs:385]: Yes, I think this needs the same treatment as the progress branch. But maybe in the next PR?

File: src/provider/mod.rs (hunk 2 above)
Comment (rklaehn) [reply src/provider/mod.rs:439]: Was meant to replace this with `num_cpus::get()`.

File: src/util.rs (hunk 3 above)
Comment (rklaehn) [reply src/util.rs:187]: It is fine if less than the buffer is being read. The validation happens in the decoder. We just want to stop when read returns 0, which means EOF.

File: src/main.rs (hunk 1 above)
Comment (rklaehn) [reply src/main.rs:394]: You got it...

image

**Comment by rklaehn** [comment]
Other than showing progress, this is ready to go.

I would do showing progress together with https://github.com/n0-computer/iroh/pull/853 since it is essentially the same thing.

File: src/provider/mod.rs (hunk 2 above)
Comment (dignifiedquire) [reply src/provider/mod.rs:439]: not replaced yet

File: src/util.rs [hunk 6]
```diff
@@ -152,6 +154,46 @@ impl From<anyhow::Error> for RpcError {
 /// A serializable result type for use in RPC responses.
 pub type RpcResult<T> = result::Result<T, RpcError>;
 
+#[derive(Debug, Display, Error)]
+pub(crate) enum BaoValidationError {
+    /// Generic io error. We were unable to read the data.
+    IoError(io::Error),
+    // /// The hash of the data does not match the hash of the outboard.
+    // HashMismatch,
```
Comment (dignifiedquire) [inline src/util.rs:162]: Why are these commented out/not yet used?

File: src/provider/database.rs [hunk 7]
```diff
@@ -271,6 +275,58 @@ impl Database {
...
     }
 
+    /// Validate the entire database, including collections.
+    ///
+    /// This works by taking a snapshot of the database, and then validating. So anything you add after this call will not be validated.
+    pub(crate) fn validate(
+        &self,
+        parallelism: usize,
+    ) -> impl Stream<Item = (Hash, u64, Option<PathBuf>, Option<BaoValidationError>)> {
+        let mut data = self
```
[the remark this answers is not in the document]
Comment (rklaehn) [reply src/provider/database.rs:285]: I think it is fine. The DB is mostly bytes::Bytes, and cloning those is basically cloning an Arc.

But will add a comment.

File: src/util.rs (hunk 6 above)
Comment (rklaehn) [reply src/util.rs:162]: I want the validation errors to be a bit more informative in the future. E.g. it should say at what offset a hash mismatch occurs. Currently it just passes through the errors from the bao validation.

But I can remove the commented out stuff if you want.

**Review by dignifiedquire — APPROVED** [review]

**Pull request merged**